product-mini/platforms/posix/main.c

```diff
@@ -934,38 +939,52 @@ main(int argc, char *argv[])
         native_lib_list, native_lib_count, native_lib_loaded_list);
 #endif
 
-    /* load WASM byte buffer from WASM bin file */
-    if (!(wasm_file_buf =
-              (uint8 *)bh_read_file_to_buffer(wasm_file, &wasm_file_size)))
+    fd = open(wasm_file, O_RDONLY, 0);
+
+    if (fd < 0) {
+        printf("Open wasm app file [%s] failed.\n", wasm_file);
         goto fail1;
+    }
 
-#if WASM_ENABLE_AOT != 0
-    if (wasm_runtime_is_xip_file(wasm_file_buf, wasm_file_size)) {
+    if (fstat(fd, &wasm_file_info) < 0) {
+        printf("Get wasm app file info [%s] failed.\n", wasm_file);
+        goto fail1;
+    }
+
+    /* mmap the wasm file */
+    if ((wasm_file_buf = mmap(NULL, wasm_file_info.st_size,
+                              PROT_READ | PROT_EXEC, MAP_PRIVATE, fd, 0))
+        == MAP_FAILED) {
+        printf("Map wasm app file [%s] failed.\n", wasm_file);
+        goto fail1;
+    }
+
+    /* only if the file XIP file, we can use the file buffer directly, for
+     * bytecode file, we need to copy the file buffer
+     * TODO: AOT file can also be used directly if new AOT format is used
```

Comment:
which new AOT format?

if we are changing the format, it would be nicer to allow simpler detection of xip.
like having a small header so that wasm_runtime_is_xip_file only needs to look at it.
(rather than mmap/read the entire file)

Reply:
Good idea.

The GC feature will comes with breaking changes in dev/gc_refactor branch including above change.